accepted/future-releases/small-features-21Q1/feature-specification.md

```diff
@@ -0,0 +1,109 @@
+
+The changes, in more detail, are detailed below.
+
+## Allow type arguments on annotations
+
+We currently allow metadata to be a call to a constant constructor:
+
+```dart
+@Deprecated("Do not use this thing")
+```
+
+However, the *grammar* does not allow type arguments, meaning that it's not possible to write:
+
+```dart
+@TypeHelper<int>(42, "The meaning")
+```
+
+There is no technical reason for this restriction. It was just simpler, and probably didn't seem necessary at the time metadata was introduced. It does now.
+
+The only change is in the grammar, from
+
+```
+<metadatum> ::= \gnewline{}
+  <identifier> | <qualifiedName> | <constructorDesignation> <arguments>
+```
+
+to
+
+```
+<metadatum> ::= \gnewline{}
+  <identifier> | <qualifiedName> | <constructorDesignation> <argumentPart>
+```
+
+The corresponding constructor invocation must still be valid, now including the type arguments.
+
+The constructed constant, if accessible in any way, will contain the provided type arguments, exactly like if it had been created by a <code>\`const' \<constructorDesignation> \<argumentPart></code> production, and is canonicalized correspondingly.
+
+The largest expected effort for this implementation is the analyzer adding a place to store the type arguments to its public AST. The remaining changes should be using existing functionality.
+
+If type arguments are allowed and omitted, the types are inferred from the types of the arguments to the constructor, as for any other constant invocation. This already happens (checked in VM with `dart:mirrors`), so no change is necessary.
+
+## Allow generic function types as type arguments
+
+The language disallows generic function types as type arguments.
+
+```dart
+List<T Function<T>(T)> idFunctions; // INVALID
+var callback = [foo<T>(T value) => value]; // Inferred as above, then invalid.
+```
+
+We remove that restriction, so a type argument *can* be a generic function type.
+
+This requires no new syntax, and in some cases only the removal of a single check. There might be some platforms where the implementation currently assumes that generic function types cannot occur as the value of type variables (an proof-of-concept attempt hit an assert in the VM). Such assumptions will need to be flushed out with tests and fixed.
+
+Because we already infer `List<T Function<T>(T)>` in the code above, this change will not affect type inference, it will just make the inferred type not be an error afterwards.
+
+We do not expect the removal of this restriction to affect the feasibility of type inference. After all, it's already possible to have a generic function type occurring covariantly in a type argument, like `List<T Function<T>(T) Function()>`.
+
+## Allow `>>>` as overridable operator
```

Comment:
This section does not specify that `>>>=` is also legal syntax. Should it? Looks like the VM currently supports it with `--enable-experiment=triple-shift`.

Reply:
It follow from being an `<operator>`, because we allow composite assignment for all `<operator>`s, but I should probably write it explicitly.

Reply:
Excellent, thanks!